### PR TITLE
Change 4461 replace hard coded user roles with fully configurable user roles

### LINKED ISSUE #4461: Replace hard-coded user roles with fully configurable user roles and remove REST_USER user role [5]
### Problem Description
The current user roles are hard-coded, but need to become manageable by the admin. There is already a first step made towards this in #830, which allows to override the user rights configuration in the database.

Source: https://github.com/hzi-braunschweig/SORMAS-Project/issues/898#issuecomment-671336289

### Proposed Change

Depends on #2804.

#### Preparation

- [x] Existing UserRole enum should become private DefaultUserRole enum and should be moved to sormas-backend
- [x] Move mapping between DefaultUserRole and UserRight from UserRight enum to DefaultUserRole enum

#### UserRole entity

- [x] The existing UserRoleConfig class needs to become the new UserRole class. When loading user roles only use the UserRole entries from the database (in other words: Don't mix-in user roles created on the fly based on the default user roles or something like that).
- [x] UserRole: Add caption and description fields
- [x] UserRole: Add hasOptionalHealthFacility, hasAssociatedOfficer and portHealthUser boolean fields. Those might be replaced with other mechanics in the future
- [x] UserRole: Add a list of notifications to the user role
- [x] User: Replace Set of old UserRole enum with Set of new UserRole entity. Will require a user:userRoles mapping table.
- [x] Provide a method to create user roles based on all default roles. Should be called similar to the default users when the system is set up
- [x] ~~Should we keep a defaultUserRole field to know that a user roles was created based on a default user role? Could be helpful for migrations later. Otherwise~~ remove the userRole field.
- [x] the getAssignableUserRoles logic can be dropped - it should actually already have been dropped with #659

#### Migration

- [x] When updating from a previous version call the method to create user roles based on the default user roles
- [x] Migration of user roles for existing users

#### Android App

- [x] Replace userRolesJson in User with join table
- [x] UserRoleConfig becomes UserRole
  - Create new table and drop the old one
  - Include all fields that have been added in #4461 to the UserRole. Notification types can be ignored
  - userRights can be stored as json
- [x] Change the UserDao.getUsersByAssociatedOfficer method (introduced by #8318) to retrieve useres based on the actual user rights (join user roles and do a "like" check on the user rights json)
- [x] Same for the logic needed for getRandomRegion/DistrictCaseResponsible introduced by #8318

Out of scope: Add translations for user role names

## Release Notes
The REST_USER default user role has been removed, because it has become obsolete after we had to remove the logic that has given users of this role access to data across all jurisdictions.
All default user roles now come with the SORMAS_REST user right, so for most use cases (sormas angular, external visits) this will not have any effect. For the survnet converter the user roles of the related user have to be changed to the NATIONAL_USER.

### PR DESCRIPTION
**Pull request by ChristopherRiedel**
Fixes #4461

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/caze/CaseExportDto.java [hunk 1]
```diff
@@ -2327,8 +2328,8 @@ public String getReportingUserRoles() {
 		return reportingUserRoles;
 	}
 
-	public void setReportingUserRoles(Set<UserRole> roles) {
-		this.reportingUserRoles = StringUtils.join(roles, ", ");
+	public void setReportingUserRoles(Set<UserRoleDto> roles) {
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/java/de/symeda/sormas/api/caze/CaseExportDto.java:2331]: Use UserRoleReferenceDto

File: sormas-api/src/main/java/de/symeda/sormas/api/event/EventCriteria.java [hunk 2]
```diff
@@ -68,7 +68,7 @@ public class EventCriteria extends CriteriaWithDateType implements ExternalShare
 	private EventInvestigationStatus eventInvestigationStatus;
 	private Disease disease;
 	private DiseaseVariant diseaseVariant;
-	private UserRole reportingUserRole;
+	private UserRoleDto reportingUserRole;
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/java/de/symeda/sormas/api/event/EventCriteria.java:71]: Use UserRoleReferenceDto

File: sormas-api/src/main/resources/enum.properties [hunk 3]
```diff
@@ -1610,60 +1610,60 @@ UserRightGroup.SYSTEM_CONFIGURATION = System Configuration
...
-UserRole.Short.POE_SUPERVISOR = POESup
-UserRole.Short.STATE_OBSERVER = RegObs
-UserRole.Short.SURVEILLANCE_SUPERVISOR = SurvSup
-UserRole.Short.ADMIN_SUPERVISOR = AdminSup
-UserRole.Short.SURVEILLANCE_OFFICER = SurvOff
-UserRole.Short.REST_EXTERNAL_VISITS_USER = ExtVis
-UserRole.Short.REST_USER = ReST
-UserRole.Short.SORMAS_TO_SORMAS_CLIENT = SormasToSormas
-UserRole.Short.BAG_USER = BAG
+# DefaultUserRole
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/resources/enum.properties:1613]: Move up based on alphabetical ordering of this file

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]
I have started by reviewing the sormas-api project.
There are multiple usages of UserRoleDto that need to be replaced by UserRoleReferenceDto.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java [hunk 4]
```diff
@@ -755,7 +768,9 @@ public UserSyncResult syncUser(String uuid) {
 	@Override
 	public List<UserDto> getUsersWithDefaultPassword() {
 		User currentUser = userService.getCurrentUser();
-		if (currentUser.getUserRoles().stream().anyMatch(r -> r.hasDefaultRight(UserRight.USER_EDIT))) {
+		if (userRoleFacade.hasUserRight(
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserFacadeEjb.java:771]: Use currentUserService.hasUserRight

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserRightsFacadeEjb.java [hunk 5]
```diff
@@ -140,8 +141,8 @@ private void generateUserRightsDocument(Map<UserRole, Set<UserRight>> userRoleRi
 		sheet.setColumnWidth(0, 256 * 35);
 		sheet.setColumnWidth(1, 256 * 50);
 		sheet.setColumnWidth(2, 256 * 75);
-		sheet.createFreezePane(2,2,2,2);
-		for (UserRole userRole : UserRole.values()) {
+		sheet.createFreezePane(2, 2, 2, 2);
+		for (DefaultUserRole userRole : DefaultUserRole.values()) {
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserRightsFacadeEjb.java:145]: Use roles from userRoleRights instead

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserRightsFacadeEjb.java [hunk 6]
```diff
@@ -157,7 +158,7 @@ private void generateUserRightsDocument(Map<UserRole, Set<UserRight>> userRoleRi
 		final Cell jurDescHeadlineCell = jurisdictionRow.createCell(1);
 		jurDescHeadlineCell.setCellValue(I18nProperties.getCaption(Captions.UserRight_jurisdictionOfRole));
 		jurDescHeadlineCell.setCellStyle(boldStyle);
-		for (UserRole userRole : UserRole.values()) {
+		for (DefaultUserRole userRole : DefaultUserRole.values()) {
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserRightsFacadeEjb.java:161]: Use roles from userRoleRights instead

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserService.java [hunk 7]
```diff
@@ -493,7 +493,7 @@ public Predicate buildCriteriaFilter(UserCriteria userCriteria, CriteriaBuilder
 			filter = CriteriaBuilderHelper.and(cb, filter, cb.equal(from.get(User.ACTIVE), userCriteria.getActive()));
 		}
 		if (userCriteria.getUserRole() != null) {
-			Join<User, UserRole> joinRoles = from.join(User.USER_ROLES, JoinType.LEFT);
+			Join<User, DefaultUserRole> joinRoles = from.join(User.USER_ROLES, JoinType.LEFT);
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserService.java:496]: Should be UserRole

File: sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserService.java [hunk 8]
```diff
@@ -660,13 +657,12 @@ public Long countByPointOfEntries(List<District> districts, UserRight userRight)
...
-	public boolean hasRole(UserRole userRoleName) {
-		return getCurrentUser().getUserRoles().contains(userRoleName);
+	public boolean hasRole(UserRole userRole) {
+		return getCurrentUser().getUserRoles().contains(userRole);
 	}
 
 	public boolean hasRight(UserRight right) {
-		User currentUser = getCurrentUser();
-		return userRoleConfigFacade.getEffectiveUserRights(currentUser.getUserRoles().toArray(new UserRole[0])).contains(right);
+		return getCurrentUser().getUserRoles().stream().flatMap(role -> role.getUserRights().stream()).collect(Collectors.toSet()).contains(right);
```
Comment (MartinWahnschaffe) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/user/UserService.java:665]: Replace with currentUserService.hasRight

File: sormas-ui/src/test/java/de/symeda/sormas/ui/TestDataCreator.java [hunk 9]
```diff
@@ -742,4 +747,27 @@ public RDP(RegionDto region, DistrictDto district, PointOfEntryDto pointOfEntry)
...
+			userRoleDto.setHasOptionalHealthFacility(defaultUserRole.hasOptionalHealthFacility());
+			userRoleDto.setEmailNotifications(defaultUserRole.getEmailNotifications());
+			userRoleDto.setSmsNotifications(defaultUserRole.getSmsNotifications());
+			userRoleDto.setJurisdictionLevel(defaultUserRole.getJurisdictionLevel());
+			FacadeProvider.getUserRoleFacade().saveUserRole(userRoleDto);
+			userRoleDtoMap.put(defaultUserRole, userRoleDto.toReference());
+		});
+	}
+
+	public Map<DefaultUserRole, UserRoleReferenceDto> getUserRoleDtoMap() {
```
Comment (MartinWahnschaffe) [inline sormas-ui/src/test/java/de/symeda/sormas/ui/TestDataCreator.java:766]: Replace with getUserRoleReference(DefaultUserRole) (see backend tests)

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]
A few changes needed for tests in backend and ui.

Final part missing of the code review is the sormas-app. Will take care of that later today.

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]

File: sormas-ui/src/main/java/de/symeda/sormas/ui/UserProvider.java [hunk 10]
```diff
@@ -43,8 +43,12 @@ public class UserProvider {
 
 	private UserDto user;
 	private UserReferenceDto userReference;
+	private Set<UserRoleDto> userRoles;
 	private Set<UserRight> userRights;
 	private JurisdictionLevel jurisdictionLevel;
+	private Boolean portHealthUser;
+	private Boolean hasAssociatedOfficer;
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/UserProvider.java:50]: See above

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/resources/captions.properties [hunk 11]
```diff
@@ -2604,15 +2604,15 @@ UserRight.description=Description
...
 UserRight.jurisdictionOfRole=Jurisdiction of role
 UserRole.userRights=User rights
-UserRole.emailNotifications=Email notifications
+UserRole.emailNotificationTypes=Email notification types
 UserRole.enabled=Enabled
 UserRole.caption=Caption
 UserRole.description=Description
-UserRole.hasOptionalHealthFacility=Has Optional Health Facility
+UserRole.hasOptionalHealthFacility=Has optional health facility
 UserRole.hasAssociatedOfficer=Has associated officer
```
Comment (MateStrysewske) [inline sormas-api/src/main/resources/captions.properties:2612]: Translation has to be adjusted here (= "Has associated district user")

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 12]
```diff
@@ -11441,8 +11441,8 @@ ALTER TABLE userroles ADD COLUMN description varchar(4096);
 ALTER TABLE userroles_history ADD COLUMN description varchar(4096);
 ALTER TABLE userroles ADD COLUMN hasoptionalhealthfacility boolean default false;
 ALTER TABLE userroles_history ADD COLUMN hasoptionalhealthfacility boolean default false;
-ALTER TABLE userroles ADD COLUMN hasassociatedofficer boolean default false;
-ALTER TABLE userroles_history ADD COLUMN hasassociatedofficer boolean default false;
+ALTER TABLE userroles ADD COLUMN hasassociateddistrcituser boolean default false;
```
Comment (MateStrysewske) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:11444]: Typo

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 13]
```diff
@@ -11441,8 +11441,8 @@ ALTER TABLE userroles ADD COLUMN description varchar(4096);
...
+ALTER TABLE userroles_history ADD COLUMN hasassociateddistrcituser boolean default false;
```
Comment (MateStrysewske) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:11445]: Typo

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDataCreator.java [hunk 14]
```diff
@@ -155,10 +155,10 @@ private void createUserRoles() {
 			userRoleDto.setCaption(defaultUserRole.toString());
 			userRoleDto.setEnabled(true);
 			userRoleDto.setPortHealthUser(defaultUserRole.isPortHealthUser());
-			userRoleDto.setHasAssociatedOfficer(defaultUserRole.hasAssociatedOfficer());
+			userRoleDto.setHasAssociatedDistrictUser(defaultUserRole.hasAssociatedOfficer());
```
Comment (MateStrysewske) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/TestDataCreator.java:158]: Looks like the setter in `DefaultUserRole` has not been renamed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/UserProvider.java [hunk 15]
```diff
@@ -165,10 +165,10 @@ public boolean isPortHealthUser() {
 	}
 
 	public boolean hasAssociatedOfficer() {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/UserProvider.java:167]: Rename this method as well

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
Tests are failing because some renamings also need to be reflected in the mobile app

**Comment by ChristopherRiedel** [comment]
> Tests are failing because some renamings also need to be reflected in the mobile app

should be fixed

**Review by MateStrysewske — APPROVED** [review]

**Pull request merged**